File: vn/src/main/java/org/sqlbuilder/vn/VnSemanticsXmlProcessor.java
package org.sqlbuilder.vn;

import org.sqlbuilder.common.XmlProcessor;
import org.w3c.dom.Element;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

import java.io.IOException;

import javax.xml.parsers.ParserConfigurationException;

public class VnSemanticsXmlProcessor extends XmlProcessor
{
	static final char START_ARGS = '(';
	static final char END_ARGS = ')';
	static final String SEP_ARGS = ", ";
	static final String SEP_FORMULA = "\n";

	static final String EVENT = "(during|start|end|result)\\(E[01]?\\)";
	static final String SPECIFIC = "(Direction|Endstate|Emotion|Form|Instrument|\\??Material|Motion|Odor|\\??Patient|Pos|Prop|Role|Sound|\\?Theme|Theme_[ij]|Weather_type)";
	// "Direction"
	// "Endstate"
	// "Form"
	// "Material"
	// "Motion"
	// "Pos"
	// "Prep"
	// "Prop"
	static final String CONSTANT = "(abstract/physical|directedmotion|forceful|from|illegal|physical|physical/abstract|toward)";

	// "abstract"
	// "abstract/physical"
	// "directedmotion"
	// "forceful"
	// "from"
	// "illegal"
	// "physical"
	// "physical/abstract"
	// "toward"

	public VnSemanticsXmlProcessor()
	{
		//
	}

	@Override
	public String process(final String xml) throws ParserConfigurationException, SAXException, IOException
	{
		if (xml == null || xml.isEmpty())
			return null;
		final Element e = XmlProcessor.docFromString("<root>" + xml + "</root>");
		return VnSemanticsXmlProcessor.process(e);
	}

	private static String process(final Element root) throws RuntimeException
	{
		root.normalize();

		final StringBuilder sb = new StringBuilder();

		final NodeList preds = root.getElementsByTagName("PRED");
		final int m = preds.getLength();
		for (int i = 0; i < m; i++)
		{
			final Element pred = (Element) preds.item(i);

			final String neg = pred.getAttribute("bool");
			if (!neg.isEmpty())
			{
				sb.append(neg);
			}

			final String value = pred.getAttribute("value");
			sb.append(value);

			// args
			final NodeList argss = pred.getElementsByTagName("ARGS");
			if (argss.getLength() > 1)
				throw new RuntimeException("ARGS number at pred " + i);

			// args
			sb.append(VnSemanticsXmlProcessor.START_ARGS);
			int a = 0;
			final NodeList args = pred.getElementsByTagName("ARG");
			final int n = args.getLength();
			for (int j = 0; j < n; j++)
			{
				final Element arg = (Element) args.item(j);
				final String t = arg.getAttribute("type");
				final String v = arg.getAttribute("value");

				if (t.equalsIgnoreCase("ThemRole"))
				{
					if (a++ > 0)
					{
						sb.append(VnSemanticsXmlProcessor.SEP_ARGS);
					}
					sb.append(v);
				}
			}
			for (int j = 0; j < n; j++)
			{
				final Element arg = (Element) args.item(j);
				final String t = arg.getAttribute("type");
				final String v = arg.getAttribute("value");

				if (!t.equalsIgnoreCase("ThemRole"))
				{
					if (a++ > 0)
					{
						sb.append(VnSemanticsXmlProcessor.SEP_ARGS);
					}
					if (t.equalsIgnoreCase("Event"))
					{
						if (!v.matches("E[01]?") && !v.matches(VnSemanticsXmlProcessor.EVENT))
							throw new RuntimeException("Event expression " + v);
						sb.append("event:").append(v);
					}
					else if (t.equalsIgnoreCase("VerbSpecific"))
					{
						if (!v.matches(VnSemanticsXmlProcessor.SPECIFIC))
							throw new RuntimeException("VerbSpecific expression " + v);
						sb.append("verbspecific:").append(v);
					}
					else if (t.equalsIgnoreCase("Constant"))
					{
						if (!v.matches(VnSemanticsXmlProcessor.CONSTANT))
							throw new RuntimeException("Constant expression " + v);
						sb.append("constant:").append(v);
					}
					else
						throw new RuntimeException("type " + t);
				}
			}
			sb.append(VnSemanticsXmlProcessor.END_ARGS);
			if (i < m - 1)
			{
				sb.append(VnSemanticsXmlProcessor.SEP_FORMULA);
			}
		}
		return sb.toString();
	}
}

File: vn/src/main/java/org/semantikos/vn/collector/VnSemanticsXmlProcessor.kt
package org.semantikos.vn.collector

import org.semantikos.common.Utils.camelCase
import org.semantikos.common.XmlProcessor
import org.w3c.dom.Element
import org.xml.sax.SAXException
import java.io.IOException
import java.util.regex.Pattern
import javax.xml.parsers.ParserConfigurationException

class VnSemanticsXmlProcessor : XmlProcessor() {

    @Throws(ParserConfigurationException::class, SAXException::class, IOException::class)
    override fun process(xml: String): String {
        val e = docFromString("<root>$xml</root>")
        return process(e)
    }

    companion object {

        private const val LOG_ONLY = false
        private const val TARGET_34 = true

        const val START_ARGS: Char = '('
        const val END_ARGS: Char = ')'
        const val SEP_ARGS: String = ", "
        const val SEP_FORMULA: String = "\n"

        const val EVENT_REGEX33: String = "E[0-1]?"
        const val EVENT_REGEX34: String = "[eëE][0-8]?"
        val EVENT_ASPECT_REGEX: String = "(during|end|result|start)\\(" + (if (TARGET_34) EVENT_REGEX34 else EVENT_REGEX33) + "\\)"

        const val SPECIFIC_REGEX33: String = "Direction|Emotion|Endstate|Form|Instrument|Light|Material|Motion|Odor|Patient|Pos|Prep|Prop|Role|Sound|Theme|Theme_i|Theme_j|Weather_type"
        const val SPECIFIC_REGEX34: String =
            "V_Attribute|V_Configuration|V_Destination|V_Direction|V_Emotion|V_Eventuality|V_Final_State|V_Form|V_Injury|V_Instrument|V_Light|V_Material|V_Manner|V_Movement|V_Odor|V_Orientation|V_Patient|V_Position|V_Spatial_Relation|V_Sound|V_State|V_Theme|V_Trajectory|V_Weather"

        const val CONSTANT_REGEX33: String = "ch_of_loc|ch_of_poss|ch_of_state|ch_on_scale|deceptive|directedmotion|forceful|hostile|illegal|playful|quality|toward|tr_of_info"
        const val CONSTANT_REGEX34: String =
            "abstract|agent|manner|directedmotion|fast|forceful|from|illegal|importance|movement|physical|physical/abstract|toward|ch_of_loc|ch_of_poss|ch_of_state|ch_on_scale|tr_of_info|stimulated|stimulating|enticed|enticing|deceptive|deceived|hostile|playful|unserious|quality|negative_emotion|slow"

        val EVENT_PATTERN: Pattern = Pattern.compile(if (TARGET_34) EVENT_REGEX34 else EVENT_REGEX33, Pattern.CASE_INSENSITIVE)
        val EVENT_ASPECT_PATTERN: Pattern = Pattern.compile(EVENT_ASPECT_REGEX, Pattern.CASE_INSENSITIVE)
        val SPECIFIC_PATTERN: Pattern = Pattern.compile(if (TARGET_34) SPECIFIC_REGEX34 else SPECIFIC_REGEX33, Pattern.CASE_INSENSITIVE)
        val CONSTANT_PATTERN: Pattern = Pattern.compile(if (TARGET_34) CONSTANT_REGEX34 else CONSTANT_REGEX33, Pattern.CASE_INSENSITIVE)

        @Throws(RuntimeException::class)
        private fun process(root: Element): String {
            root.normalize()

            val sb = StringBuilder()

            val preds = root.getElementsByTagName("PRED")
            val m = preds.length
            for (i in 0..<m) {
                val pred = preds.item(i) as Element

                val neg = pred.getAttribute("bool").trim { it <= ' ' }
                if (!neg.isEmpty()) {
                    sb.append(neg)
                }

                val value = pred.getAttribute("value").trim { it <= ' ' }
                sb.append(value)

                // args
                val argss = pred.getElementsByTagName("ARGS")
                if (argss.length > 1) {
                    throw RuntimeException("ARGS number at pred $i")
                }

                // args
                sb.append(START_ARGS)
                var a = 0
                val args = pred.getElementsByTagName("ARG")
                val n = args.length
                for (j in 0..<n) {
                    val arg = args.item(j) as Element
                    val t = arg.getAttribute("type").trim { it <= ' ' }
                    val v = arg.getAttribute("value").trim { it <= ' ' }

                    if (t.equals("ThemRole", ignoreCase = true)) {
                        if (a++ > 0) {
                            sb.append(SEP_ARGS)
                        }
                        sb.append(v)
                    }
                }
                for (j in 0..<n) {
                    val arg = args.item(j) as Element
                    val t = arg.getAttribute("type").trim { it <= ' ' }
                    var v = arg.getAttribute("value").trim { it <= ' ' }
                    if (!t.equals("ThemRole", ignoreCase = true)) {
                        if (a++ > 0) {
                            sb.append(SEP_ARGS)
                        }

                        // EVENT
                        if (t.equals("Event", ignoreCase = true)) {
                            val v2 = if (v[0] == '?') v.substring(1) else v
                            // System.out.println("@EVENT " + v);
                            if (EVENT_PATTERN.matcher(v2).matches()) {
                                v = v.uppercase()
                            } else if (EVENT_ASPECT_PATTERN.matcher(v2).matches()) {
                                v = camelCase(v)
                            } else {
                                if (LOG_ONLY) {
                                    System.err.println("SEMANTICS Event expression <$v>")
                                } else {
                                    throw RuntimeException("SEMANTICS Event expression <$v>")
                                }
                            }
                            sb.append("event:").append(v)
                        } else if (t.equals("VerbSpecific", ignoreCase = true)) {
                            val v2 = if (v[0] == '?') v.substring(1) else v
                            //System.out.println("@SPEC " + v);
                            if (!SPECIFIC_PATTERN.matcher(v2).matches()) {
                                if (LOG_ONLY) {
                                    System.err.println("SEMANTICS VerbSpecific expression <$v>")
                                } else {
                                    throw RuntimeException("SEMANTICS VerbSpecific expression <$v>")
                                }
                            }
                            sb.append("verbspecific:").append(v)
                        } else if (t.equals("Constant", ignoreCase = true)) {
                            //System.out.println("@CONST " + v);
                            if (!CONSTANT_PATTERN.matcher(v).matches()) {
                                if (LOG_ONLY) {
                                    System.err.println("SEMANTICS Constant expression <$v>")
                                } else {
                                    throw RuntimeException("SEMANTICS Constant expression <$v>")
                                }
                            }
                            sb.append("constant:").append(v)
                        } else if (t.equals("PredSpecific", ignoreCase = true)) {
                            // ignore
                        } else {
                            throw RuntimeException("type $t")
                        }
                    }
                }
                sb.append(END_ARGS)
                if (i < m - 1) {
                    sb.append(SEP_FORMULA)
                }
            }
            return sb.toString()
        }
    }
}
